// @metaplex-foundation/js/dist/esm/plugins/walletAdapterIdentity/WalletAdapterIdentityDriver.mjs
import { UninitializedWalletAdapterError, OperationNotSupportedByWalletAdapterError } from '../../errors/SdkError.mjs';

class WalletAdapterIdentityDriver {
  constructor(walletAdapter) {
    this.walletAdapter = walletAdapter;
  }
  get publicKey() {
    if (!this.walletAdapter.publicKey) {
      throw new UninitializedWalletAdapterError();
    }
    return this.walletAdapter.publicKey;
  }
  async signMessage(message) {
    if (this.walletAdapter.signMessage === undefined) {
      throw new OperationNotSupportedByWalletAdapterError('signMessage');
    }
    return this.walletAdapter.signMessage(message);
  }
  async signTransaction(transaction) {
    if (this.walletAdapter.signTransaction === undefined) {
      throw new OperationNotSupportedByWalletAdapterError('signTransaction');
    }
    return this.walletAdapter.signTransaction(transaction);
  }
  async signAllTransactions(transactions) {
    if (this.walletAdapter.signAllTransactions === undefined) {
      throw new OperationNotSupportedByWalletAdapterError('signAllTransactions');
    }
    return this.walletAdapter.signAllTransactions(transactions);
  }
}

export { WalletAdapterIdentityDriver };
